### PR TITLE
fix(ui-markdown-editor): resolve hyperlink overflow - I219

### LINKED ISSUE #219: Hyperlink modal goes off page
# Bug Report 🐛
The hyperlink modal sometimes goes off the page if the link is near the edge of the page.
image

I may provide a mockup for how this could be fixed later, but open to suggestions

### PR DESCRIPTION
**Pull request by K-Kumar-01**
# Closes #219 <CORRESPONDING ISSUE NUMBER>

### Changes

- <ONE> Changed the index.js and HyperlinkModal.js in ui-markdown-editor
- <TWO> Styling changes done in above files

### Flags

- <ONE> Had to style the components according to the screen size by hit and trial and observing different breakpoints.(brute force method)
- <TWO> Had to reduce modal width to ensure it fits (screens less than 335px have not been adjusted)

### Screenshots or Video

ss2
ss1

### Related Issues
- Issue #219 <NUMBER>

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [ ] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [ ] Merging to `master` from `fork:branchname`
- [ ] Manual accessibility test performed
    - [ ] Keyboard-only access, including forms
    - [ ] Contrast at least WCAG Level A
    - [ ] Appropriate labels, alt text, and instructions

### CODE REVIEW

**Comment by Michael-Grover** [comment]
Looks great in the screenshots you sent! 

I tried it out in Netlify and it looks like the arrow isn't pointing to the highlighted area. I'm highlighting where it says "I'm highlighting this" in this screenshot

image

**Comment by K-Kumar-01** [comment]
> Looks great in the screenshots you sent!
> 
> I tried it out in Netlify and it looks like the arrow isn't pointing to the highlighted area. I'm highlighting where it says "I'm highlighting this" in this screenshot
> 
> image

On my laptop, no such problem occurred as I only changed the left position of the arrow.
If possible could you tell me the screen size on which the following problem occurred, it would be helpful.
Also, netlify link would really help me to test it actually on small screens and not on developer tools.
Thanks :)

**Comment by Michael-Grover** [comment]
I was viewing at 1440px screen width on a 15" Macbook pro using Google Chrome

You can find the [Netlify preview](https://deploy-preview-222--ap-web-components.netlify.app/) by clicking the bottom right "Details" button here:

image

**Comment by K-Kumar-01** [comment]
DeepinScreenshot_20201211011547
I am using brave on deepin
Screen size is "1536px".
The arrow still seems to be correct on my laptop.
I also checked on Google Chrome resulting in the same.
I think that it might have something to do with os we are using. Not sure, though.

**Comment by Michael-Grover** [comment]
If you have a chance can you send a screenshot of the arrow pointing somewhere other than "link"? What happens if you try to make "this is the text" a hyperlink, does the arrow move to point there?

**Comment by K-Kumar-01** [comment]
> If you have a chance can you send a screenshot of the arrow pointing somewhere other than "link"? What happens if you try to make "this is the text" a hyperlink, does the arrow move to point there?

DeepinScreenshot_20201211115419
@Michael-Grover 
I tried making "this is the text a hyperlink". However, the arrow won't change.
The reason which I understand is that the arrow styling is hardcoded for the area at which the link is called i.e. "link" one.
So even if the link is changed arrow won't change. 
Can't we simply remove the arrow?

**Comment by irmerk** [comment]
This code is a little hard to follow, but it looks like you have made the modal stay to the left of the screen permanently, and you have changed the arrow pointer. This results in it _appearing_ to be hovering under the original "link" in the markdown editor. 

It looks like `window.innerWidth>660` is generally always true, which means `el.style.left='25px';` I don't think this is the right approach. 

The approach should not be around the window width and rather the relational position of the highlighted text/link to the border of the markdown editor. 
- If it is **less** than the width of the modal in distance from the left of the markdown editor, the position of the modal should skew to the right and it should appear **to the right** of the highlighted text/link. 
- If it is **more** than the width of the modal in distance from the left of the markdown editor, the position of the modal should skew to the left and it should appear **to the left** of the highlighted text/link.

**Comment by K-Kumar-01** [comment]
> This code is a little hard to follow, but it looks like you have made the modal stay to the left of the screen permanently, and you have changed the arrow pointer. This results in it _appearing_ to be hovering under the original "link" in the markdown editor.
> 
> It looks like `window.innerWidth>660` is generally always true, which means `el.style.left='25px';` I don't think this is the right approach.
> 
> The approach should not be around the window width and rather the relational position of the highlighted text/link to the border of the markdown editor.
> 
> * If it is **less** than the width of the modal in distance from the left of the markdown editor, the position of the modal should skew to the right and it should appear **to the right** of the highlighted text/link.
> * If it is **more** than the width of the modal in distance from the left of the markdown editor, the position of the modal should skew to the left and it should appear **to the left** of the highlighted text/link.

Okay, I will try this approach.
However, as far as I understand the arrow issue will still remain in this case also, as the text wraps also so it may cause problems. I will try my best to solve using this approach.

>It looks like window.innerWidth>660 is generally always true, which means el.style.left='25px';

Also as of this, it is true for most of the time but also changes on resize as seen here
ss2

**Comment by K-Kumar-01** [comment]
@Michael-Grover @irmerk 
```
el.style.left = `${rect.left
        + window.pageXOffset
        - el.offsetWidth / 2
        + rect.width / 2}px`;
```
Here when i **console.log(window.pageXOffset)** and **console.log(rect.width)** both give the value zero.
So is there any particular reason for them to be included or should i remove them as they don't help us in any calculatios.

**Comment by irmerk** [comment]
> ```
> el.style.left = `${rect.left
>         + window.pageXOffset
>         - el.offsetWidth / 2
>         + rect.width / 2}px`;
> ```
> 
> Here when i **console.log(window.pageXOffset)** and **console.log(rect.width)** both give the value zero.
> So is there any particular reason for them to be included or should i remove them as they don't help us in any calculatios.

I think any of this can be rewritten however solves this overall issue.

**Comment by K-Kumar-01** [comment]
@irmerk 
I have updated the PR, making the rendering of the modal position and arrow dynamic.
I had to reduce the modal width as it would have caused problems otherwise in some situations(namely 660 to 600px and screens less than 440px).
Please let me know if there are any changes required. :)

DeepinScreenshot_20201211235427
DeepinScreenshot_20201212001358
DeepinScreenshot_20201212001419

**Comment by K-Kumar-01** [comment]
@irmerk @Michael-Grover 
Any updates on my PR?

**Review by DianaLease — APPROVED** [review]
This looks good to me! I'd like to get @Michael-Grover & @irmerk 's feedback as well since they had originally looked at this PR before the latest changes.

File: packages/ui-markdown-editor/src/FormattingToolbar/index.js [hunk 1]
```diff
@@ -54,11 +54,36 @@ const FormattingToolbar = ({
...
+      if (rightEndPos > containerWidth) {
+        let diff = rightEndPos-containerWidth;
+        // extra space of 10px on right side to look clean
+        diff+=10;
+        calPos=calPos-diff;
+        let shift=diff-5;
+        w.style.left= `calc(50% + ${shift}px)`;
+      }
+
+      el.style.left = `${calPos}px`;
```
Comment (irmerk) [inline packages/ui-markdown-editor/src/FormattingToolbar/index.js:86]: This looks great!!

Some semantic nitpicks:
- Could you name `w` in `let w = el.children[0];` more descriptively or describe in a comment?
- `z = el.parentElement;` doesn't seem necessary, as later you could just do:
  `let containerWidth = el.parentElement.offsetWidth;`
- Could you name `calPos` in `let calPos = rect.left  - el.offsetWidth / 2;` more descriptively or describe in a comment?
- Some of these variables don't change and thus can be `const`:
`const rightEndPos = calPos + el.offsetWidth;`
`const containerWidth = z.offsetWidth;`

**Review by Michael-Grover — APPROVED** [review]
lgtm, thanks!

File: packages/ui-markdown-editor/src/FormattingToolbar/index.js (hunk 1 above)
Comment (K-Kumar-01) [reply packages/ui-markdown-editor/src/FormattingToolbar/index.js]: @irmerk 
I have made a PR with the above suggestions implemented.
Please let me know if there are any changes required :)

**Review by irmerk — APPROVED** [review]
Amazing! 🎉
